Mark JUnit as test scope; upgrade JUnit to 4.13; upgrade Traitor to 1.4.0

iamjpotts opened the pull request:
The JUnit upgrade deprecates `ExpectedException` and `org.junit.Assert.assertThat`

A read-only internal collection broke Traitor 1.3.0, but that was already fixed in 1.4.0 so a Traitor upgrade is also included.

jnape commented on pom.xml:
```diff
@@ -63,6 +63,8 @@
             <groupId>junit</groupId>
             <artifactId>junit</artifactId>
+            <version>4.13</version>
+            <scope>test</scope>
```
`junit` is a managed dependency from the `com.jnape.palatable.parent` pom, where its scope is indeed specified as `test`, which is why you don't see it here.  So this scope is redundant.

jnape replied:
I'd probably rather upgrade the dependency in the parent and let it cascade down. I need to think about this change because to make sure there are no ramifications to the test-support types that the lambda `test-jar` offers for client libraries that use the older junit versions.

iamjpotts replied:
Good catch; I'm going to do some more digging also.

iamjpotts commented:
Here's what I found in my analysis of the impact of a JUnit and Mockito upgrade in the parent pom and the projects that use it. No animals were harmed and social distancing was maintained.

Parent pom

junit: replace 4.11 with 4.13
mockito: replace mockito-all 2.3.2 with mockito-core 3.3.3

Changes with JUnit 4.13:
Deprecated: org.junit.Assert.assertThat
Replaced by: org.hamcrest.MatcherAssert.assertThat

Deprecated: ExpectedException rule
Replaced by: assertThrows

Deprecated: org.mockito.runners.MockitoJUnitRunner
Replaced by: org.mockito.junit.MockitoJUnitRunner


Benchmark
Replace imports of MockitoJUnitRunner

DMutext
Gradle project; doesn't use parent pom

Dynamic Collections
Saw a reference to this in Benchmark but didn't see a git repo for it

Lambda
Replace imports of assertThat
Replace usage of ExpectedException with assertThrows

Lambda-Koans
Replace imports of assertThat

Loan Shark
Replace imports of MockitoJUnitRunner

Ouroboros 
No changes

Parlance
No changes (looks empty)

Shoki
Replace imports of assertThat

Traits
Replace imports of assertThat with
Additional new mock for fireTestSuiteFinished()

Lambda is used in production apps but I'm not sure how much the other repos are used, or what your maintenance/support plans are for them. Some of them haven't seen a commit in about 6 years.

I'm interested to see what comes of Parlance.

Does a path forward look like upgrading JUnit and Mockito in the parent pom, and then only upgrading the parent pom version in the active repos? Or would we upgrade the parent pom of all of the repos? Something else?

I've got most of the changes for the various libraries locally already, so I'm happy to post up PRs for whatever you think is appropriate; I imagine that would start with the parent pom.

iamjpotts commented:
Postponing until a later major upgrade.
